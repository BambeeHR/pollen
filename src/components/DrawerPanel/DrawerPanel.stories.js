import { select } from '@storybook/addon-knobs';
import BaseButton from '../BaseButton/BaseButton.vue';
import DrawerPanel from './DrawerPanel.vue';
import Modal from '../../constants/Modal';
import UserAvatar from '../UserAvatar/UserAvatar.vue';

export default {
  title: 'DrawerPanel',
  component: DrawerPanel,
  parameters: {
    componentSubtitle: '',
  },
};

export const Default = () => ({
  components: { BaseButton, DrawerPanel, UserAvatar },
  props: {
    size: {
      default: select('Size', Object.values(Modal.Sizes), Modal.Sizes.REGULAR),
    },
  },
  data() {
    return {
      image: 'https://cdn.filestackcontent.com/enCa5BTVeIsLHcgkdpAc',
      isActive: true,
    };
  },
  template: `
  <div>
    <BaseButton @click="isActive = !isActive">Toggle Panel</BaseButton>
    <DrawerPanel :active="isActive" class="test-class" @close="isActive = false" :size="size">
      <template #header>
        <UserAvatar :image="image" name="Xavier Oaxaca" title="HR Manager" class="ml-5" />
      </template>
      <div class="px-5 pb-6">
        <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Mauris fermentum ornare porttitor. Orci varius natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Mauris eu massa id nulla varius vulputate. Etiam sed semper neque. Maecenas a mauris sit amet erat iaculis euismod pulvinar non turpis. Curabitur ultrices nibh at nisi molestie sollicitudin. Phasellus ullamcorper feugiat maximus. Vestibulum pellentesque turpis tellus, iaculis maximus justo mollis at. Vestibulum et est lectus. In eleifend ac lectus sollicitudin finibus. Duis dolor lorem, suscipit a mattis in, vestibulum ac diam.</p>
        <p>Pellentesque et mattis neque, id aliquet tortor. Praesent congue suscipit lectus. Mauris cursus in justo ac ullamcorper. Quisque quis leo eu justo condimentum consequat. Interdum et malesuada fames ac ante ipsum primis in faucibus. Maecenas eu elementum nisi. Nullam sed nisi elementum, lacinia justo sed, dapibus odio. Donec maximus eros ac tortor tristique consequat. Nunc cursus est arcu. Curabitur turpis nibh, consectetur at lectus vel, efficitur tempor felis.</p>
        <p>Sed posuere tortor dolor, non malesuada velit sodales non. Aliquam dictum mauris libero, sed fermentum elit accumsan sit amet. Nam dictum urna eu tortor cursus, at ultricies eros luctus. Nulla imperdiet elementum aliquet. Aliquam justo orci, ultricies quis purus vitae, aliquet mattis nisl. Vivamus id accumsan ex, vitae laoreet ligula. Morbi est orci, suscipit ac neque a, ultrices malesuada justo.</p>
        <p>Pellentesque maximus neque risus, in aliquam felis molestie non. Nunc egestas odio metus, vitae vestibulum leo rhoncus eu. Sed congue dapibus nunc vel semper. Nulla eget diam ante. Donec finibus lacinia suscipit. In commodo laoreet magna non convallis. Nullam sit amet leo non nunc mattis mattis. Pellentesque sollicitudin ipsum vel sollicitudin luctus. Pellentesque blandit ex vel dapibus euismod. Suspendisse fringilla dictum dui, vitae pulvinar odio. Suspendisse at leo vitae neque fermentum feugiat. Vestibulum fringilla sapien id gravida cursus. Sed sagittis pulvinar sollicitudin. Interdum et malesuada fames ac ante ipsum primis in faucibus. Integer dictum sem sit amet odio venenatis malesuada.</p>
        <p>Maecenas a dictum velit, a mattis dui. Praesent nec tellus ut purus porttitor vestibulum vitae vel libero. In nulla diam, fringilla eget faucibus ac, vestibulum id orci. Curabitur pulvinar porttitor dictum. Nunc tristique ante ut odio congue sollicitudin. In id condimentum tortor, eu fringilla nibh. Maecenas condimentum metus ac rhoncus bibendum. Maecenas iaculis purus ipsum, ac malesuada elit hendrerit id. Vestibulum consequat lectus ac ligula volutpat interdum.</p>
      </div>
      <div class="bg-gray-6 sticky bottom-0 mt-auto px-5 py-3 text-center">
        <BaseButton @click="isActive = false" type="button">Close</BaseButton>
      </div>
    </DrawerPanel>
  </div>
  `,
});
