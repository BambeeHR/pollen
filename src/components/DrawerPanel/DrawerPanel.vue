<template>
  <FocusLock :disabled="!active" class="focus-lock">
    <div
      class="drawer-panel"
      :class="[`drawer-panel--${size}`, { 'drawer-panel--active': active }]"
      :aria-hidden="!active"
    >
      <header v-if="!!$slots.header" class="drawer-panel__header">
        <slot name="header" />
      </header>
      <slot />
      <ModalClose v-if="showClose && active" variant="light" @click="close" />
    </div>
    <div v-if="active" class="drawer-panel__overlay" @click="close" />
  </FocusLock>
</template>

<script>
import FocusLock from 'vue-focus-lock';
import ModalClose from '../internal/modal/ModalClose.vue';
import Modal from '../../constants/Modal';

export default {
  components: { FocusLock, ModalClose },
  inheritAttrs: false,
  props: {
    /**  If true, makes the drawer panel visible and locks focus within the panel. */
    active: {
      type: Boolean,
      default: false,
    },
    /** If true, displays the close button in the top right of the panel. */
    showClose: {
      type: Boolean,
      default: true,
    },
    /**
     * The overlay width. One of `tiny`, `small`, `regular`, `medium`, `large,
     * or `giant`.
     */
    size: {
      type: String,
      default: Modal.Sizes.REGULAR,
      validator: (value) => Object.values(Modal.Sizes).includes(value),
    },
  },
  watch: {
    active: {
      async handler(newValue) {
        await this.$nextTick();
        if (newValue) {
          document.addEventListener('keyup', this.handleKeyup);
        } else {
          document.removeEventListener('keyup', this.handleKeyup);
        }
      },
      immediate: true,
    },
  },
  beforeDestroy() {
    document.removeEventListener('keyup', this.handleKeyup);
  },
  methods: {
    /**
     * Close the modal if the user presses the escape key.
     * @param {Event} e
     */
    handleKeyup(e) {
      if (e.keyCode !== 27) {
        return;
      }
      this.close();
    },
    close() {
      this.$emit('close');
    },
  },
};
</script>

<style scoped>
.drawer-panel {
  @apply bg-white duration-light fixed flex flex-col inset-y-0 overflow-y-auto right-0 shadow-4 transform translate-x-full w-screen z-50;
}

.drawer-panel--active {
  @apply translate-x-0;
}

.drawer-panel__header {
  @apply border-b border-gray-4 border-solid flex flex-shrink-0 items-center pr-12 border-t-0 border-r-0 border-l-0;
  min-height: 4.5rem;
}

.drawer-panel__overlay {
  @apply fixed inset-0 bg-transparent z-40;
}
.drawer-panel--tiny {
  max-width: 19.5rem;
}

.drawer-panel--small {
  max-width: 25rem;
}

.drawer-panel--regular {
  max-width: 30rem;
}

.drawer-panel--medium {
  max-width: 37.5rem;
}

.drawer-panel--large {
  @apply max-w-3xl;
}

.drawer-panel--giant {
  @apply max-w-5xl;
}

.drawer-panel >>> .modal-close {
  @apply absolute right-2 top-2;
}
</style>
