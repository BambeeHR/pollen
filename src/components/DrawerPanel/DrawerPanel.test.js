import { mount } from '@vue/test-utils';
import DrawerPanel from './DrawerPanel.vue';

describe('DrawerPanel', () => {
  test('renders correctly', () => {
    const wrapper = mount(DrawerPanel, {
      slots: {
        default: 'Content goes here',
        header: 'Heading goes here',
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  test('emits a close event when escape is pressed', async () => {
    const elem = document.createElement('div');
    if (document.body) {
      document.body.appendChild(elem);
    }
    const wrapper = mount(DrawerPanel, {
      attachTo: elem,
      propsData: { active: true },
    });
    await wrapper.vm.$nextTick();
    wrapper.trigger('keyup.enter');
    expect(wrapper.emitted().close).toBeFalsy();
    wrapper.trigger('keyup.esc');
    expect(wrapper.emitted().close.length).toBe(1);
    wrapper.destroy();
    wrapper.trigger('keyup.esc');
    expect(wrapper.emitted().close.length).toBe(1);
  });

  test('emits a close event when the overlay is clicked', async () => {
    const wrapper = mount(DrawerPanel, {
      propsData: { active: true },
    });
    await wrapper.vm.$nextTick;
    wrapper.find('.drawer-panel__overlay').trigger('click');
    expect(wrapper.emitted().close.length).toBe(1);
  });
});
