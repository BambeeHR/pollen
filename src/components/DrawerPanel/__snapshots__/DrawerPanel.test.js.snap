// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrawerPanel renders correctly 1`] = `
<div
  class="focus-lock"
>
  <div
    aria-hidden="true"
    tabindex="-1"
  />
   
  <div
    aria-hidden="true"
    tabindex="-1"
  />
   
  <div
    data-lock=""
  >
    <div
      aria-hidden="true"
      class="drawer-panel drawer-panel--regular"
    >
      <header
        class="drawer-panel__header"
      >
        Heading goes here
      </header>
       Content goes here 
      <!---->
    </div>
     
    <!---->
  </div>
   
  <div
    aria-hidden="true"
    tabindex="-1"
  />
</div>
`;
